heartbeat/service_test.go

```diff
@@ -0,0 +1,273 @@
+		waiters: make(map[basics.Round]chan struct{}),
+		history: []table{nil}, // some genesis accounts could go here
+		hdr: bookkeeping.BlockHeader{
+			UpgradeState: bookkeeping.UpgradeState{
+				CurrentProtocol: protocol.ConsensusFuture,
+			},
+		},
+	}
+}
+
+func (l *mockedLedger) LastRound() basics.Round {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.lastRound()
+}
+func (l *mockedLedger) lastRound() basics.Round {
+	return basics.Round(len(l.history) - 1)
+}
+
+func (l *mockedLedger) WaitMem(r basics.Round) chan struct{} {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if l.waiters[r] == nil {
+		l.waiters[r] = make(chan struct{})
+	}
+
+	// Return an already-closed channel if we already have the block.
+	if r <= l.lastRound() {
+		close(l.waiters[r])
+		retChan := l.waiters[r]
+		delete(l.waiters, r)
+		return retChan
+	}
+
+	return l.waiters[r]
+}
+
+// BlockHdr allows the service access to consensus values
+func (l *mockedLedger) BlockHdr(r basics.Round) (bookkeeping.BlockHeader, error) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if r > l.lastRound() {
+		return bookkeeping.BlockHeader{}, fmt.Errorf("%d is beyond current block (%d)", r, l.LastRound())
+	}
+	// return the template hdr, with round
+	hdr := l.hdr
+	hdr.Round = r
+	return hdr, nil
+}
+
+// setSeed allows the mock to return a specific seed
+func (l *mockedLedger) setSeed(seed committee.Seed) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	l.hdr.Seed = seed
+}
+
+func (l *mockedLedger) addBlock(delta table) error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	l.history = append(l.history, delta)
+
+	for r, ch := range l.waiters {
+		switch {
+		case r < l.lastRound():
+			fmt.Printf("%d < %d\n", r, l.lastRound())
+			panic("why is there a waiter for an old block?")
+		case r == l.lastRound():
+			close(ch)
+			delete(l.waiters, r)
+		case r > l.lastRound():
+			/* waiter keeps waiting */
+		}
+	}
+	return nil
+}
+
+func (l *mockedLedger) LookupAccount(round basics.Round, addr basics.Address) (ledgercore.AccountData, basics.Round, basics.MicroAlgos, error) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if round > l.lastRound() {
+		panic("mockedLedger.LookupAccount: future round")
+	}
+
+	for r := round; r <= round; r-- {
+		if acct, ok := l.history[r][addr]; ok {
+			more := basics.MicroAlgos{Raw: acct.MicroAlgos.Raw + 1}
+			return acct, round, more, nil
+		}
+	}
+	return ledgercore.AccountData{}, round, basics.MicroAlgos{}, nil
+}
+
+// waitFor confirms that the Service made it through the last block in the
+// ledger and is waiting for the next. The Service is written such that it
+// operates properly without this sort of wait, but for testing, we often want
+// to wait so that we can confirm that the Service *didn't* do something.
+func (l *mockedLedger) waitFor(s *Service, a *require.Assertions) {
+	a.Eventually(func() bool { // delay and confirm that the service advances to wait for next block
+		_, ok := l.waiters[l.LastRound()+1]
+		return ok
+	}, time.Second, 10*time.Millisecond)
```

Comment:
NIT: Should we wait longer for the sake of slow build machines?

Reply:
I'm willing to, but I've never seen it fail.  You want 2 seconds?